### DISCUSSION TITLE
Next.js-before-hydration takes too long both in prod and dev

### DISCUSSION THREAD

**Discussion by abriginets** [vercel/next.js#29857] [Help]
### What version of Next.js are you using?

11.1.2

### What version of Node.js are you using?

14.18.0

### What browser are you using?

Chrome 94

### What operating system are you using?

Windows 10 WSL2 Ubuntu 20.04

### How are you deploying your application?

Vercel

### Describe the Bug

Hello. I'm experiencing difficulties measuring performance of my NextJS project. I mentioned that sometimes it takes up to 13 seconds for my production deployment to respond to the request. Before I start I would like to mention that I'm testing the homepage with SSR which requires some amount of data which is getting fetched from Upstash and PlanetScale (Redis and MySQL). Every single response from 3rd party API is being cached in Redis for a considerable amount of time but for some reason it still loads for 13 seconds in an hour or so after the previous request was made. I tried looking into Performance tab in Chrome DevTools but I just can't understand what does `Next.js-before-hydration` mean and what exactly it does, but obviously it takes too much time:

![image](https://user-images.githubusercontent.com/17240507/137022559-f4bcadbb-9635-4a8f-b69e-aebdd866fa0f.png)

All the data that is required to fill the page with content can be fetched at ~200-300 milliseconds (I monitored it for some time), but where does Next.JS spends 1.7 seconds more? 

This issue only occur for cold starts. If I reload the page right after my homepage loaded for the first time since the startup it'll take only ~350ms for server to respond. Is there something I could do to improve performance of my project's cold start page loading time?

### Expected Behavior

NextJS shouldn't spend that much time serving the page

### To Reproduce

-

**Comment by ijjk** [comment]
Hi, are you able to provide more information a repo with a minimal reproduction maybe? Also how much data is being returned from `getServerSideProps`/`getStaticProps`? How are you confirming how long the data fetching is taking, are you measuring inside of `getServerSideProps` or externally?

**Reply by abriginets** [reply]
Hello. Unfortunately I can't provide you with minimal reproduction, my code is copyrighted.

> how much data is being returned from getServerSideProps

It's 34kb of json (noncompressed)

> How are you confirming how long the data fetching is taking

By using `console.time()` around the main call:
```ts
console.time('index props');
const pageProps = await instance.getProps(userIp);
console.timeEnd('index props'); // outputs index props: 524.475ms
```

If I FLUSHALL my Redis cache instance it'll display around 2-3 seconds, but NextJS will serve the page in 4-5 seconds. If the load time is ~500ms then NextJS will serve the page in ~2,5 seconds.

**Reply by ijjk** [reply]
It sounds like there is some logic inside of your page component's react tree causing a slow down in rendering, you can use tools like mentioned [here](https://reactjs.org/docs/perf.html) or [here](https://reactjs.org/blog/2018/09/10/introducing-the-react-profiler.html#profiling-an-application) to narrow in what could be causing this.

**Reply by abriginets** [reply]
It appears that the most time consuming task is `serialize` which is the part of Webpack (https://github.com/sevenCon/blog-github/blob/d7da03ca50c00bada83322f089dd58148238a9ab/demo/webpack/webpack-master/lib/serialization/PlainObjectSerializer.js):
![image](https://user-images.githubusercontent.com/17240507/137197706-b13c3f8b-0a8f-4b80-958c-77ca208d4cea.png)

**Reply by abriginets** [reply]
It seems like Webpack is trying to serialize something really big on the backend for the initial startup. The following page reloads appears to be fast as expected.

**Reply by ijjk** [reply]
Hmm that doesn't look like a trace from the react profiler, can you try profiling there and make sure to run against development so the proper display names are shown?

**Reply by abriginets** [reply]
Yea, this is the backend profiler in Chrome. I'm using the following command to start Next locally in dev mode: `NODE_OPTIONS='--inspect' next dev` and then I see green NodeJS logo in DevTools clicking on which I can open dedicated DevTools allowing me to profile backend side of project. Profiling client side really doesn't make much sense to me because issue is obviously at the backend. Frontend seems to be performing very fast:
![image](https://user-images.githubusercontent.com/17240507/137225444-32a417bd-8628-40eb-8648-4ac1d4e41f17.png)

**Reply by ijjk** [reply]
If the problem was in your page's component tree you should see the same performance drop on the client.  From the above info it looks like that is not the case. Can you share any snippets from `getServerSideProps` or the page being rendered? Are you doing any data fetching in `_app` or elsewhere?

**Reply by abriginets** [reply]
That's the code I execute to get all the necessary props for homepage:

```ts
  async getProps(userIp: string): Promise<BaseIndexPageProps> {
    const location = await this.travelpayoutsService.getLocation(userIp);

    const [popularDirections, specialOffers, currencies, currencyRates, countries, cities] = await Promise.all([
      this.travelpayoutsService.getPopularDirectionsFromCity({ origin: location.data.iata, currency: 'RUB' }),
      this.travelpayoutsService.getSpecialOffersForIndexPage(),
      this.databaseService.getCurrencyList(AllowedCurrencyLanguagesEnum.RU),
      this.travelpayoutsService.getCurrencyRates(),
      this.databaseService.findCountriesForFooter(),
      this.databaseService.findCitiesForFooter(),
    ]);

    return {
      cities,
      countries,
      currencies,
      currencyRates,
      specialOffers,
      popularDirections,
      userGeolocation: location.data,
    };
  }
```

Every single method from `travelpayoutsService` uses caching with Redis. Upstash dashboard shows that longest response from Redis took ~197ms. Everything from `databaseService` makes requests to PlanetScale MySQL and their dashboard shows that the slowest query for the last week took 87ms.

So I decided to run a cron that continiously makes GET requests to my Vercel deployment (every 5 minutes with `?_vercel_no_cache=1` to keep lambda container warm and avoid hitting it cold) and here's the results from not so long ago:
![image](https://user-images.githubusercontent.com/17240507/137543614-4c4da1d9-ce03-486e-a07c-ec38b2ff3f6d.png)

In avg my backend performs pretty good but then there's this 12 seconds page load time. Weird. I've set up timing functionality with the following decorator and started monitoring:
```ts
export function Timing() {
  return function (target: unknown, name: string, descriptor: PropertyDescriptor): void {
    const method = descriptor.value;

    descriptor.value = async function () {
      const startTime = new Date(Date.now());      
      const result = await method.apply(this);
      const endTime = new Date(Date.now());
      const elpased = endTime.getTime() - startTime.getTime();

      // Write the elpased time to the new function's metadata.
      Reflect.defineMetadata(LAST_CALL_ELAPSED_KEY, elpased, descriptor.value);
      
      return result;
    };
  };
}
```
And my backend was fetching the props for homepage normally:
```
X-Backend-Timing: 975 // miliseconds from `@Timing` decorator, represents the time elapsed on calling and executing `getProps` method
Date: Fri, 15 Oct 2021 17:58:17 GMT
X-Vercel-Cache: BYPASS
```
And also I've checked my Logflare report about this request:
```
    "report": {
      "billed_duration_ms": 11943,
      "duration_ms": 11943,
      "init_duration_ms": 382,
      "max_memory_used_mb": 331,
      "memory_size_mb": 1024
    },
```
So if lambda took ~400ms to init and my backend took ~975ms to fetch all the data it needs then my backend should've responded in around 2 seconds tops considering the network latencies and stuff. What could possibly be elapsing almost 10 seconds more? I don't have an answer. Am I doing something wrong with benchmarking and monitoring?

**Reply by abriginets** [reply]
@ijjk Is there any progress on this investigation? I've been struggling with making Next.JS startup times good enough to stop seeing bad TTFB warning from Google PageSpeed. It seems like amount of content or React components I'm using makes no difference and the issue comes from Next.JS itself. It that possible right now to merge https://github.com/vercel/next.js/pull/23261 for testing purposes? I see that the author proposes to make it experimental. If that's my issue then we could make it as a standard for SSR pages.

**Reply by abriginets** [reply]
UPD: if I remove a component that uses a bunch of redux-toolkit reducers SSR takes 4 seconds instead of 12. Why does redux eats SSR performance so hard that I don't know yet...
